Feature/test suspend

xross opened the pull request:
~#138 should be addressed first.~

Note, requires associated changes in xsim.

Also addresses some XS2 related build issues (relating to vbus changes)

DaveAtkinsonXMOS reviewed (commented):
Let me know what you think of the comments...

xross commented on examples/AN00135_test_and_measurement_class/host/usbtmc_test.py:
```diff
@@ -5,27 +5,28 @@
 # Find XMOS USBTMC test device
-dev = usb.core.find(idVendor=0x20b1, idProduct=0x2337)
+dev = usb.core.find(idVendor=0x20B1, idProduct=0x2337)
 
 
 import usbtmc
-instr =  usbtmc.Instrument(0x20b1, 0x2337)
+
+instr = usbtmc.Instrument(0x20B1, 0x2337)
```
I ran all the python through the auto-formatting tool "black" (I should have done this on a separate PR..)
I've made no functional changes to any of the app notes

xross commented on lib_xud/src/core/XUD_Main.xc:
```diff
@@ -288,23 +289,24 @@ static int XUD_Manager_loop(XUD_chan epChans0[], XUD_chan epChans[],  chanend ?c
                 if (one)
                 {
-                    reset = XUD_Init();
+                    #if defined(XUD_SIM_XSIM) || defined(XUD_BYPASS_CONNECT) 
+                        reset = 1;
+                    #else
+                        reset = XUD_Init();
+                    #endif
```
I don't normally indent either, so fixed it. Extra indent was a mistake - we should be using an auto formatter ideally.

DaveAtkinsonXMOS replied:
Cool...

xross commented on tests/helpers.py:
```diff
@@ -6,181 +6,265 @@
-        sim_args += ['--vcd-tracing', vcd_args]
-#        sim_args += ['--xscope', '-offline logs/xscope.xmt']
+        sim_args += ["--vcd-tracing", vcd_args]
+    #        sim_args += ['--xscope', '-offline logs/xscope.xmt']
```
This was a formatting only change but I removed the dead code.

xross commented on tests/helpers.py:
```diff
@@ -6,181 +6,265 @@
-    """ Choose the size of a frame near the minimum size frame (46 data bytes)
-    """
+    """Choose the size of a frame near the minimum size frame (46 data bytes)"""
     return rand.randint(46, 54)
```
Dead code, removed.

xross commented on tests/usb_event.py:
```diff
@@ -0,0 +1,33 @@
+import abc
+
+
+class UsbEvent(object):
```
Hangover from py2. Fixed.

xross commented on tests/usb_packet.py:
```diff
@@ -58,359 +58,548 @@
+            interEventDelay=interEventDelay,
+        )
+
+    return sofToken
```
Fixed

xross commented on tests/usb_packet.py:
```diff
@@ -58,359 +58,548 @@
 
 def GenCrc16(args):
 
     data = args
-  
-    crc = 0xffff;
-    poly = 0x8005;
 
-    for byte in  data:
-        topBit = 1 << 15;
-        crc ^= reflect(int(byte) & int(0xFF), 8) << 8; 
+    crc = 0xFFFF
+    poly = 0x8005
 
-        for k in range(0,8):
+    for byte in data:
+        topBit = 1 << 15
+        crc ^= reflect(int(byte) & int(0xFF), 8) << 8
+
+        for k in range(0, 8):
             if crc & topBit:
-                crc = (crc << 1) ^ poly;
+                crc = (crc << 1) ^ poly
             else:
-                crc <<= 1;
+                crc <<= 1
+
+    # //crc = crc ^0xffff;
+    crc = reflect(crc, 16)
+    crc = ~crc
+    crc = crc & 0xFFFF
+    # print "CRC: : {0:#x}".format(crc)
+    return crc
 
-    #//crc = crc ^0xffff;
-    crc = reflect(crc,16);
-    crc = ~crc;
-    crc = crc & 0xffff;
-    #print "CRC: : {0:#x}".format(crc)
-    return crc;
```
From memory I think the this was a direct translation from the reference in the spec. Doesn't hurt to speed it up since we're now using it more in anger (Note, largest packet size is 1k)

Also note, this was a formatting only change.

xross replied:
@DaveAtkinsonXMOS  This is causing some test fails. I've not debugged, will revert for now.

xross commented on tests/usb_packet.py:
```diff
@@ -58,359 +58,548 @@
 
     while iBitcnt > 0:
-        if ((udata ^ crc5) & (0x1 << (intSize - 1))):   #bit4 != bit4?
-            crc5 <<= 1;
-            crc5 ^= poly5;
+        if (udata ^ crc5) & (0x1 << (intSize - 1)):  # bit4 != bit4?
+            crc5 <<= 1
+            crc5 ^= poly5
         else:
-            crc5 <<= 1;
-        udata <<= 1;
-        iBitcnt = iBitcnt-1
+            crc5 <<= 1
+        udata <<= 1
+        iBitcnt = iBitcnt - 1
 
-    #Shift back into position
-    crc5 >>= intSize - 5;
+    # Shift back into position
+    crc5 >>= intSize - 5
 
-    #Invert contents to generate crc field
-    crc5 ^= 0x1f;
-    
-    crc5 = reflect(crc5, 5);
-    return crc5;
+    # Invert contents to generate crc field
+    crc5 ^= 0x1F
+
+    crc5 = reflect(crc5, 5)
+    return crc5
```
Done

xross commented on tests/usb_phy_shim.py:
```diff
@@ -7,34 +7,63 @@
-            #error_nibbles = packet.get_error_nibbles()
-            
+        for i, packet in enumerate(self._packets):
+            # error_nibbles = packet.get_error_nibbles()
```
Done

xross commented on tests/usb_phy_shim.py:
```diff
@@ -45,79 +74,84 @@ def run(self):
                     timeout = timeout - 1
-                    #print "{i}".format(i=timeout)
+                    # print "{i}".format(i=timeout)
 
-                    #sample TXV for new packet
+                    # sample TXV for new packet
```
Done

xross commented on tests/usb_phy_shim.py:
```diff
@@ -45,79 +74,84 @@ def run(self):
 
-                #if isinstance(packet, TokenPacket):
-                 #   print "Token packet, clear valid token"
+                # if isinstance(packet, TokenPacket):
```
Done

xross commented on tests/usb_phy_shim.py:
```diff
@@ -146,12 +180,12 @@ def run(self):
                             print("ERROR: Unexpected packet from xCORE")
 
-                    #print "Sending byte {0:#x}".format(byte)
+                    # print "Sending byte {0:#x}".format(byte)
```
Removed

xross commented on tests/usb_phy_shim.py:
```diff
@@ -166,22 +200,19 @@ def run(self):
 
-                #if self._verbose:
-                    #print "Sent"
+                # if self._verbose:
```
done

xross commented on tests/usb_signalling.py:
```diff
@@ -0,0 +1,178 @@
+
+                fsTime_ns = xsi.get_time()
+                timeToFs_ns = fsTime_ns - suspendStartTime_ns
+                # print("DEVICE ENTERED FS AT TIME " + str(fsTime_ns/1000) + "(after " + str(timeToFs_ns/1000) +" uS)")
```
Done

xross commented on tests/usb_signalling.py:
```diff
@@ -0,0 +1,178 @@
+
+            time_ns = xsi.get_time() - suspendStartTime_ns
+            if time_ns == self._duration_ns:
+                # print("SUSPEND END: " + str(xsi.get_time()))
```
Done

xross commented on tests/runtests.py:
```diff
@@ -73,8 +93,9 @@ def epilogue(args):
 #    * Outputting with timestamps
-#""")
-#'''
+#""",
```
This was an update from black

xross reviewed (commented):
@DaveAtkinsonXMOS, I *think* I have addressed all comments. Thanks for the review.

mbanth commented on tests/test_suspend_resume/LICENSE.rst:
```diff
@@ -0,0 +1,84 @@
+*******************************
```
On 2021-05-05, @lucianomartin removed the license files from the test directories as part of the preparation for making the repo public.  You can see this change in commit c8a7bd8.  I asked him to do so because the tests are not distributed separately from the repo, so there's no need for an additional license file and having duplicates increases the risk of them diverging.  Please do not reintroduce per-test license files.

xross replied:
Copy pasta. Agree'd should not be duplicated.

mbanth commented on tests/test_suspend_resume/src/main.xc:
```diff
@@ -0,0 +1,68 @@
+// Copyright 2016-2021 XMOS LIMITED.
```
It looks like this is a new file.  If so, the copyright date should be just '2021'.

xross replied:
copy pasta. fixed

mbanth commented on tests/test_suspend_resume/src/xc_ptr.h:
```diff
@@ -0,0 +1,48 @@
+// Copyright 2016-2021 XMOS LIMITED.
```
If this is a new file, the copyright date should be just '2021'.

xross replied:
Rm'ed file

mbanth commented on tests/usb_signalling.py:
```diff
@@ -0,0 +1,178 @@
+from usb_event import UsbEvent
```
Copyright and license reference comments missing.  Should be in every source file including test source files.

mbanth commented on lib_xud/src/core/XUD_HAL.xc:
```diff
@@ -148,12 +147,10 @@ void XUD_HAL_EnterMode_PeripheralFullSpeed()
 void XUD_HAL_EnterMode_PeripheralChirp()
 {
-#ifndef XUD_SIM_XSIM
 #ifdef __XS3A__
```
In files that assume that no definition of `__XS3A__` assumes definition of `__XS2A__`, I would prefer to see near the top of the file:
```
#if !defined (__XS2A__) && !defined (__XS3A__)
#error Unknown architecture
#endif
```
That way the compiler complains when moving to new architectures rather than silently mixing together combinations of XS2 and XS3 functionality.

mbanth reviewed (changes requested):
I have made a few comments, but generally the changes look good.  However, I did not see an update to XUD_DeviceAttach.xc which is where I see the unprotected call to `write_periph_word()`.

xross commented on tests/test_makefile.mak:
```diff
@@ -20,18 +20,20 @@ APP_NAME =
 			   -DXUD_TEST_SPEED_HS=1 \
 			   -DXUD_BYPASS_RESET=1 \
-			   -DXUD_STARTUP_ADDRESS=1
+			   -DXUD_STARTUP_ADDRESS=1 
```
Rm'd

xross commented on lib_xud/src/core/XUD_TimingDefines.h:
```diff
@@ -4,34 +4,31 @@
 #endif
 
+#ifndef SUSPEND_T_WTWRSTHS_us
+#define SUSPEND_T_WTWRSTHS_us       200     // 200us Time beforechecking for J after asserting XcvrSelect and Termselect: T_WTRSTHS: 100-875us
```
Fixed

xross commented on lib_xud/src/core/XUD_TimingDefines.h:
```diff
@@ -4,34 +4,31 @@
-#define T_SIGATT_ULPI               (T_SIGATT_ULPI_us * REF_CLK_FREQ)
-#define T_ATTDB_us                  1000000  // 1000ms
-#define T_ATTDB                     (T_ATTDB_us * REF_CLK_FREQ)
 #define T_UCHEND_T_UCH_us           1000000  // 1000ms
```
Agreed. fixed a bunch of these in a previous PR. Fixed.

xross commented:
> I have made a few comments, but generally the changes look good. However, I did not see an update to XUD_DeviceAttach.xc which is where I see the unprotected call to `write_periph_word()`.

Thanks. 

The changes that break the build we added in the VBUS detect PR. Current tests are all mostly packet based and bypass this code...

I've fixed this up now by moving this to the HAL. I also spotted that disconnect on XS3 on invalid VBUS was also broken - it was essentially running an XS2 register write regardless of architecture. I've tested the build on the HID application note (AN00129), which of course uses the device attach code. 

Coincidentally I'm just looking at the authoring the device attach tests at the moment.

xross commented on tests/usb_event.py:
```diff
@@ -1,4 +1,5 @@
-from abc import ABC, abstractmethod, abstractproperty
+# Copyright 2016-2021 XMOS LIMITED.
+# This Software is subject to the terms of the XMOS Public Licence: Version 1.rom abc import ABC, abstractmethod, abstractproperty
```
How odd, I wonder if this is Blacks doing...

xross replied:
Fixed

xross commented on tests/usb_session.py:
```diff
@@ -1,4 +1,5 @@
-from usb_phy import USB_DATA_VALID_COUNT
+# Copyright 2021 XMOS LIMITED.
+# This Software is subject to the terms of the XMOS Public Licence: Version 1.rom usb_phy import USB_DATA_VALID_COUNT
```
Fixed

xross commented on tests/usb_transaction.py:
```diff
@@ -1,4 +1,5 @@
-from usb_event import UsbEvent
+# Copyright 2021 XMOS LIMITED.
+# This Software is subject to the terms of the XMOS Public Licence: Version 1.rom usb_event import UsbEvent
```
fixed

xross commented on lib_xud/src/core/XUD_HAL.xc:
```diff
@@ -230,6 +229,39 @@ void XUD_HAL_EnterMode_PeripheralTestJTestK()
+{
+#ifdef __XS2A__
+    write_periph_word(USB_TILE_REF, XS1_SU_PER_UIFM_CHANEND_NUM, XS1_SU_PER_UIFM_FUNC_CONTROL_NUM, 4);
+#else
```
I guess there is no correct answer to this, but my current preferred approach is as follows:

Somewhere up high in the repo full check of archs and error if no valid one defined, much as you say:
#if !(defined __XS2A__) && !(defined __XS3A__) #error Unknown architecture! #endif

Everywhere else, special cases for earlier archs, then "else" current arch. This is because the next arch is likely similar to the current arch, and I've had to  spend quite a while adding #if new_arch to multiple files was a big pain.

mbanth replied:
That sounds like a good approach to me.

mbanth reviewed (approved):
Looks good
